detect/directory.go

```diff
@@ -145,3 +114,49 @@ func (d *Detector) DetectFiles(paths <-chan sources.ScanTarget) ([]report.Findin
 
 	return d.findings, nil
 }
+
+// readUntilSafeBoundary consumes |f| until it finds two consecutive `\n` characters, up to |maxPeekSize|.
+// This hopefully avoids splitting. (https://github.com/gitleaks/gitleaks/issues/1651)
+//
+// TODO: Check if the chunk ends in consecutive newlines before reading more of the file.
+func readUntilSafeBoundary(r *bufio.Reader, n int, maxPeekSize int, peekBuf *bytes.Buffer) error {
+	newlineCount := 0 // Tracks consecutive newlines
+	for {
+		data := peekBuf.Bytes()
+		if len(data) == 0 {
+			break
+		}
+
+		// Check if the last character is a newline.
+		lastChar := data[len(data)-1]
+		if lastChar == '\n' {
+			newlineCount++
+
+			// Stop if two consecutive newlines are found
+			if newlineCount >= 2 {
+				break
+			}
+		} else if lastChar == '\r' || lastChar == ' ' || lastChar == '\t' {
+			// The presence of other whitespace characters (`\r`, ` `, `\t`) shouldn't reset the count.
+			// (Intentionally do nothing.)
```

Comment:
This handles https://github.com/gitleaks/gitleaks/pull/1665#issuecomment-2539410608